### PR TITLE
Enabling `Memcached::Error` instances to bubble up to caller

### PR DESCRIPTION
**Pull request by alexsnaps**
This is a proposal on letting the callsite decide on whether exceptions raised should be silenced or should bubble up. The aim is being able to use `MemcachedStore` as a proper _store_, i.e. one that doesn't hide failures from the callsite. 
This is only step 1: I miss tests (I need to feed an actual `::Memcached` instance in there (as `Memcached::Rails` also swallows `Memcached::Error` instances. 
Then there needs to be an agreement on how to deal with `Memcached::NotFound` (i.e. misses): either leave it to the call site, or include this in `MemcachedStore` and simply return `nil` on misses.

This is more of a heads up and ground for discussion

### CODE REVIEW

**Comment by rafaelfranca** [comment]
I like the approach and I agree that we should let the callers decide what to do with the exceptions.

> Then there needs to be an agreement on how to deal with Memcached::NotFound (i.e. misses): either leave it to the call site, or include this in MemcachedStore and simply return nil on misses.

I believe in this case we should shallow the exception. In the first place a cache miss should not raise an exception even in the client library since it is not an exceptional case but a very common thing to happen.

**Comment by alexsnaps** [comment]
>I believe in this case we should shallow the exception.

That's pretty much my position so far as well. If no one thinks that's too crazy, that's how I'll proceed then

File: lib/active_support/cache/memcached_store.rb [hunk 1]
```diff
@@ -170,43 +166,47 @@ def exist?(*)
...
+          true
+        end
       end
 
       private
 
+      def handle_exceptions(default_value_on_exception:)
+        yield
+      rescue *NONFATAL_EXCEPTIONS => e
+        raise e unless @deal_with_exceptions
```
Comment (sirupsen) [inline lib/active_support/cache/memcached_store.rb]: You'll want to just do `raise` to preserve the backtrace 📝

File: lib/active_support/cache/memcached_store.rb [hunk 2]
```diff
@@ -44,6 +44,8 @@ def initialize(*addresses)
           @data = Memcached::Rails.new(*(addresses + [mem_cache_options]))
         end
 
+        @deal_with_exceptions = options.fetch(:deal_with_exceptions, true)
```
Comment (sirupsen) [inline lib/active_support/cache/memcached_store.rb]: I would suggest a name like `swallow_exceptions` or `treat_exception_as_miss`—I'd find that a little bit more descriptive when using this

**Comment by alexsnaps** [comment]
@rafaelfranca @sirupsen Merged #32 - this PR now makes more sense. 
Let me know what you guys think…

File: lib/active_support/cache/memcached_store.rb [hunk 3]
```diff
@@ -13,11 +13,14 @@ module Cache
...
       ESCAPE_KEY_CHARS = /[\x00-\x20%\x7F-\xFF]/n
 
-      attr_accessor :read_only
+      attr_accessor :read_only, :swallow_exceptions
 
       def initialize(*addresses)
         addresses = addresses.flatten
         options = addresses.extract_options!
+        @swallow_exceptions = true
+        @swallow_exceptions = options.delete(:swallow_exceptions) if options.key?(:swallow_exceptions)
```
Comment (fw42) [inline lib/active_support/cache/memcached_store.rb]: the `... if options.key?(:swallow_exceptions)` is unnecessary, isn't it? (if you remove that, the code would behave exactly the same way, wouldn't it?).

File: lib/active_support/cache/memcached_store.rb [hunk 4]
```diff
@@ -266,8 +244,14 @@ def expiration(options)
...
       end
 
-      def log_exception(e)
+      def handle_exceptions(return_value_on_error:, on_miss: return_value_on_error)
+        yield
+      rescue Memcached::NotFound, Memcached::ConnectionDataExists
+        on_miss
+      rescue Memcached::Error => e
+        raise unless @swallow_exceptions
         logger.warn("memcached error: #{e.class}: #{e.message}") if logger
```
Comment (fw42) [inline lib/active_support/cache/memcached_store.rb]: This is now not logging anymore if `@swallow_exceptions` is true, but previously it was logging I think. Is that change intended?

File: lib/active_support/cache/memcached_store.rb [hunk 5]
```diff
@@ -55,40 +58,34 @@ def read_multi(*names)
...
+              raw_values.each do |key, value|
+                entry = deserialize_entry(value)
+                values[keys_to_names[key]] = entry.value unless entry.expired?
+              end
             end
           end
+          values
         end
-        values
-      rescue Memcached::NotFound
```
Comment (fw42) [inline lib/active_support/cache/memcached_store.rb]: This was previously only rescuing NotFound, but now it's also rescueing ConnectionDataExists. Probably not a big deal. Is it?

File: test/test_memcached_store.rb [hunk 6]
```diff
@@ -594,6 +594,71 @@ def test_logger_defaults_to_rails_logger
     assert_equal Rails.logger, @cache.logger
   end
 
+  def test_constructor_set_swallow_exceptions
```
Comment (fw42) [inline test/test_memcached_store.rb]: `s/set/sets/`?

File: test/test_memcached_store.rb [hunk 7]
```diff
@@ -617,11 +682,27 @@ def with_read_only(client)
     client.read_only = previous
   end
 
+  def raises_when_not_swallowing_exceptions
```
Comment (fw42) [inline test/test_memcached_store.rb]: I think `assert_raises_when_not_swallowing_exceptions` would be a tiny bit clearer (add an `assert_` prefix to the method name).

File: lib/active_support/cache/memcached_store.rb (hunk 4 above)
Comment (fw42) [inline lib/active_support/cache/memcached_store.rb]: Nvm, misunderstood this. it's still logging if it's true and not logging (but raising) if it's false. That seems reasonable.

File: lib/active_support/cache/memcached_store.rb (hunk 3 above)
Comment (alexsnaps) [inline lib/active_support/cache/memcached_store.rb]: `@swallow_exceptions` would then be `nil`, wouldn't it? And since we raise `unless @swallow_exceptions` I think it would be different. Or is it too early in the morning for me still?

Comment (fw42) [inline lib/active_support/cache/memcached_store.rb]: you are right. sorry.

File: lib/active_support/cache/memcached_store.rb (hunk 5 above)
Comment (alexsnaps) [inline lib/active_support/cache/memcached_store.rb]: `ConnectionDataExists` is for `add` (i.e. only add if not already there) & `cas` , I could parameterize the `handle_exceptions` for additional exceptions to be treated as "miss" though (I need to deal with `ConnectionDataExists` for cas operations. Now these two cases will (or should, i.e. that's the intent) behave exactly as they did when we swallow exceptions... I thought of adding a complete "treat these exceptions as misses" kinda param to the constructor (vs. the `:swallow_exceptions`), but it felt over the top... Willing to change my mind though.

File: lib/active_support/cache/memcached_store.rb (hunk 4 above)
Comment (alexsnaps) [inline lib/active_support/cache/memcached_store.rb]: :+1:

File: test/test_memcached_store.rb (hunk 6 above)
Comment (alexsnaps) [inline test/test_memcached_store.rb]: :+1:

File: test/test_memcached_store.rb (hunk 7 above)
Comment (alexsnaps) [inline test/test_memcached_store.rb]: :+1:

File: lib/active_support/cache/memcached_store.rb [hunk 8]
```diff
@@ -266,8 +244,14 @@ def expiration(options)
         expires_in
       end
 
-      def log_exception(e)
+      def handle_exceptions(return_value_on_error:, on_miss: return_value_on_error)
```
Comment (csfrancis) [inline lib/active_support/cache/memcached_store.rb]: Nitpick: it would be nice if the keyword argument names were consistent: `on_error`, `on_miss` or `return_value_on_error`, `return_value_on_miss`.

Comment (csfrancis) [inline lib/active_support/cache/memcached_store.rb]: It's definitely not worth bike shedding about. I don't mind if you leave it as-is.

**Comment by alexsnaps** [comment]
@rafaelfranca Should we do a release for this? Or just use the specific sha?

**Comment by rafaelfranca** [comment]
Let's go with 1.1.0. Could you do the bump?
